### PR TITLE
planner: Fix the issue where any(nil) is not considered equal to nil

### LINKED ISSUE #57326: any(nil) is not equal to nil, should cast it as its original type
## Bug Report

Please answer these questions before submitting your issue. Thanks!
```
func (col *Column) Equals(other any) bool {
	if other == nil {. // false
		return false
	}
        col2, ok := other.(*Column)
        // ok will be true
        // then col2 == nil is true 
```
when we pass `var s *Column = nil` into as other with type any here, we found it doesn't equal to nil anymore, unless we downcast it with `*Column` type before equaling with nil

### 2. What did you expect to see? (Required)
as mentioned 
suggestion way
```
func (col *Column) Equals(other any) bool {
    col2, ok := other.(*Column)
    if !ok {
        return false
    }
    if col == nil {
        return col2==nil
    }
    if col2 == nil {
        return false
    }
    // normal equals logic appended
```

### 3. What did you see instead (Required)
as mentioned 

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by King-Dylan**
### What problem does this PR solve?

Issue Number: close #57326 

Problem Summary:
Fix the issue where any(nil) is not considered equal to nil.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/column.go [hunk 1]
```diff
@@ -511,9 +511,13 @@ func (col *Column) Hash64(h base.Hasher) {
 
 // Equals implements HashEquals.<1st> interface.
 func (col *Column) Equals(other any) bool {
-	if other == nil {
-		return false
+	// If col itself is nil, check whether other is also nil
+	if col == nil {
```
Comment (AilinKid) [inline pkg/expression/column.go:515]: better use this format
```
	op2, ok := other.(*DataSource)
	if !ok {
		return false
	}
	if op == nil {
		return op2 == nil
	}
	if op2 == nil {
		return false
	}
```

and there many other places implements this Equals interface with the same bad way, we should change them too

Comment (AilinKid) [reply pkg/expression/column.go:515]: image

you can do it by batch with 2-3, or finish them all, it's up to you, do remember the test case

File: pkg/expression/column.go [hunk 2]
```diff
@@ -511,9 +511,13 @@ func (col *Column) Hash64(h base.Hasher) {
...
+		// Checking if other is of type *Column and nil
+		col2, ok := other.(*Column)
+		return ok && col2 == nil
 	}
+	// Attempt to convert other to *Column
 	var col2 *Column
 	switch x := other.(type) {
```
Comment (AilinKid) [inline pkg/expression/column.go:522]: this kind of way can be eliminated, we can make it only receive *pointer as its argument here. Let the caller to pass &XX if it's a structure.

**Review by AilinKid — APPROVED** [review]
Well done

**Review by windtalker — APPROVED** [review]
lgtm

**Pull request merged**